### ISSUE TITLE
Autoconfigured unix crosstool should respect standard environment variables

### ISSUE THREAD

**Issue by hlopko** [bazelbuild/bazel#5186]
Autoconfigured C++ toolchain doesn't follow standard [gnu environment variables](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html). There is no reason why we couldn't use the standard, therefore let's do it.

Bazel currently inspects **some** environment variables, but those don't have standard naming (even within bazel), are hard to discover (you have to go read the source code) and are completely undocumented.

Since gnu env vars are widely understood, using them would lower the entry barrier plus would make the life of package maintainers simpler, as they would not have to special case packages built by bazel.

This is very simple to implement, the cost is mostly the need to migrate existing users.

The most important environment variables are:
* Programs:
	* CC: the C compiler (not C++), eg x86_64-pc-linux-gnu-gcc
	* CPP: The C pre-processor (also not C++) eg, x86_64-pc-linux-gnu-gcc -E
	* CXX: C++ compiler, eg x86_64-pc-linux-gnu-g++
	* FC: Fortran compiler
	* LD: the linker, eg x86_64-pc-linux-gnu-ld
	
Flags for the programs:
* CFLAGS,CXXFLAGS: these are passed right after $CC for any optimizations and
	        whatnot, make's implicit compile rule basically does:
	        $(CC) $(CFLAGS) foo.c -o foo.o. You are more than welcome to
	        append to any of these variables just overwriting is problematic.
	
* LDFLAGS: These are the flags for the linker itself.
* LDLIBS: these are the libraries that the binary needs to be linked
	        against. Basically stuff with -l goes in LDLIBS and stuff with
	        -L goes in LDFLAGS. make's rules for linking are roughly:
	        $(LD) $(LDFLAGS) -o out foo.o bar.o $(LDLIBS). mixing up LDFLAGS
	        and LDLIBS causes issues when --as-needed is used too.
* SYSROOT: this is for cross-compiling and autoconf passes it as gcc
	        --sysroot= i think. I don't cross-compile that often but can get
	        you more info if it'd help.

**Comment by mhlopko** [comment]
This is currently blocked by:

* #5187 - new skylark api to the C++ toolchain.
* #4571 - migration of bazel crosstools to features and action configs (needed to separate C and C++). Parts of this issue could be implemented before #4571 is fixed, but to avoid duplicated work I'd rather wait.

**Comment by rongjiecomputer** [comment]
I think there is no need to restrict this to Unix only. CMake on Windows will honour `CC`, `CXX`, `LD` and other environment variables so that user don't need to pass `-DCMAKE_C_COMPILER`, `-DCMAKE_LINKER`. flags in configure step. GCC/Mingw64 and GCC/MSYS users should generally be familiar with these environment variables. It is a nice-to-have feature for MSVC users too since these environment variables are harmless.

I am willing to implement the Windows part later. I have some other improvements to upstream after C++ Crosstool migration is done.

**Comment by mhlopko** [comment]
Good points. We have to bring our windows and darwin-with-Xcode toolchains to the same level of configurability as unix toolchain is.

**Comment by jasal82** [comment]
Can you tell me when this is going to be implemented? I'm currently evaluating Bazel as replacement for our existing CMake setup but the setup of cross toolchains is way too complicated. We could generate the necessary files from our Yocto environment but as there is already a environment setup script written by Yocto containing the CC/CXX/etc. flags it would be good if Bazel could work with that instead.

**Comment by perfinion** [comment]
@jasal82 I'm not sure for the timing, its got a few deps that arn't done yet i think. In the mean time, I do this, feel free to copy. Its kind of ugly but it works decently well. https://github.com/gentoo/gentoo/blob/0d7adf51d80763d2ca8c3e5ffac4dd9b3edbe388/sci-libs/tensorflow/tensorflow-1.11.0_rc0.ebuild#L125-L141

**Comment by jasal82** [comment]
@perfinion Your sample code is setting the compiler flags but I don't see where the cross compiler binaries are configured. Maybe I'm missing something.

**Comment by detailyang** [comment]
+1 please
